Elements/src/Spatial/AdaptiveGrid/AdaptiveGrid.cs

```diff
@@ -1347,24 +1443,30 @@ private HashSet<Edge> AddFromGrid(Grid2d grid, IEnumerable<Edge> edgesToIntersec
                 }
             };
 
-            foreach (var cell in cells)
+            var uvPolygon = (Polygon)boundingPolygon.Transformed(toGrid);
+
+            var newSegments = new List<(Vector3 from, Vector3 to)>();
+            var intersectionPoints = new List<Vector3>();
+
+            foreach (var u in uList)
             {
-                foreach (var cellGeometry in cell.GetTrimmedCellGeometry())
-                {
-                    var polygon = (Polygon)cellGeometry;
-                    for (int i = 0; i < polygon.Vertices.Count - 1; i++)
-                    {
-                        add(polygon.Vertices[i], polygon.Vertices[i + 1]);
-                    }
-                    add(polygon.Vertices.Last(), polygon.Vertices.First());
-                }
+                var verticalLine = new Line(new Vector3(u, vList.First()), new Vector3(u, vList.Last()));
+                var currentInternalSegments = verticalLine.Trim(uvPolygon, out _, includeCoincidenceAtEdge: true);
+                newSegments.AddRange(splitSegmentsWithPoints(currentInternalSegments, u, vList, false, intersectionPoints));
             }
-
-            foreach (var edge in edgeCandidates)
+            foreach (var v in vList)
             {
-                addedEdges.Add(AddInsertEdge(edge.Item1, edge.Item2));
+                var horizontalLine = new Line(new Vector3(uList.First(), v), new Vector3(uList.Last(), v));
+                var currentInternalSegments = horizontalLine.Trim(uvPolygon, out _, includeCoincidenceAtEdge: true);
+                newSegments.AddRange(splitSegmentsWithPoints(currentInternalSegments, v, uList, true, intersectionPoints));
             }
 
+            newSegments.AddRange(uvPolygon.Edges().ToList().SelectMany(e => splitSegmentWithPoints(e, intersectionPoints.Where(p => new Line(e.from, e.to).PointOnLine(p)).ToList())));
+
+            newSegments.ForEach(s => add(s.from, s.to));
```

Comment:
What if instead of adding vertices here, from segments, we do this instead  in splitSegmentsWithPoints and return pairs of vertices instead of points. Can we do less calls if just use  previousPoint/nextPoint/lastPoint but like vertices?